feat(ivy): implement `compileComponents` method for `TestBedRender3`

ocombe opened the pull request:
The implementation of the `compileComponents` method for `TestBedRender3` was missing.
We now pass each component through `resolveComponentResources` when `TestBed.compileComponents` is called so that `templateUrl` and `styleUrls` can be resolved asynchronously and used once `TestBed.createComponent` is called.
The component's metadata are overriden in `TestBed` instead of mutating the original metadata like this is the case outside of TestBed. The reason for that is that we need to ensure that we didn't mutate anything so that the following tests can run with the same original metadata, otherwise we it could trigger or hide some errors.

FW-553 #resolve

mary-poppins commented:
You can preview 75c26c1 at https://pr27778-75c26c1.ngbuilds.io/.

mary-poppins commented:
You can preview 582ee4b at https://pr27778-582ee4b.ngbuilds.io/.

mary-poppins commented:
You can preview 51b1400 at https://pr27778-51b1400.ngbuilds.io/.

mary-poppins commented:
You can preview df3e9ba at https://pr27778-df3e9ba.ngbuilds.io/.

mary-poppins commented:
You can preview af39bdf at https://pr27778-af39bdf.ngbuilds.io/.

ocombe commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -42,15 +42,9 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
-          throw new Error(error.join('\n'));
+          throw new Error(
+              `This test module uses the component ${stringify(type)} which is using a "templateUrl" or "styleUrls", but they were never compiled. ` +
+              `Please call "TestBed.compileComponents" before your test.`);
```
I think that the render3 error message was better, but if we expect it to output the same error than the view engine, then we need to use the old message.
If we don't care, then I'll revert this and change the tests to support both messages.

mhevery replied:
Please revert this and change the test to expect different msg depending on if `ivyEnabled` is on or off.

ocombe commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -259,6 +267,7 @@ export class TestBedRender3 implements Injector, TestBed {
     if (config.providers) {
       this._providerOverrides.push(...config.providers);
     }
+    this._compilerOptions.push(config);
```
I'm not sure if we need to do that here, it was how it worked in the view engine testbed, but if I omit this line then all the tests still pass.

ocombe commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -446,8 +485,19 @@ export class TestBedRender3 implements Injector, TestBed {
     class DynamicTestModule {
     }
 
+    this._createCompiler();
```
I'm not sure if we need to do that here, it was how it worked in the view engine testbed, but if I omit this line then all the tests still pass.

mhevery reviewed (commented):
Please revert the error message to the old one as it was better. LGTM otherwise

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -278,9 +287,39 @@ export class TestBedRender3 implements Injector, TestBed {
-    // TODO(pk): plug into the ivy's resource fetching pipeline
-    return Promise.resolve();
+    const resolvers = this._getResolvers();
+    const declarations: Type<any>[] = flatten(this._declarations || EMPTY_ARRAY);
```
It looks like we also need to resolve forward refs while running flatten. I have similar cases in another PR: https://github.com/angular/angular/pull/27737/files#diff-06d226868c03eb17a7d58aaa05b3d75dR495.

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -278,9 +287,39 @@ export class TestBedRender3 implements Injector, TestBed {
+      componentOverrides.forEach((override: [Type<any>, Component]) => {
+        // Once resolved, we override the existing metadata, ensuring that the resolved resources
+        // are only available until the next TestBed reset (when `resetTestingModule` is called)
+        this.overrideComponent(override[0], {set: override[1]});
```
We should probably reset/cleanup using the method described in FW-872 (quote from the ticket):
```
1) save the previous Ivy definition(s) on the classes being compiled, before it replaces them.
2) restore those definitions when `resetTestingModule` is called.
```
or this code solves different problem?

ocombe replied:
yes I think that your changes would work, I'll update my PR once yours is merged

AndrewKushnir reviewed (changes requested):
@ocombe overall looks good, I just left a couple comments/questions. Please let me know your opinion. Thank you.

ocombe commented:
PR blocked by https://github.com/angular/angular/pull/27786 (we need some of this code for that PR)

mary-poppins commented:
You can preview 98f0ef7 at https://pr27778-98f0ef7.ngbuilds.io/.

kara commented:
@ocombe This should be unblocked now that the two PRs you linked have been merged. Can you rebase?

mary-poppins commented:
You can preview e52c92a at https://pr27778-e52c92a.ngbuilds.io/.

mary-poppins commented:
You can preview 4a77b06 at https://pr27778-4a77b06.ngbuilds.io/.

mary-poppins commented:
You can preview 0be3d6e at https://pr27778-0be3d6e.ngbuilds.io/.

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -309,9 +319,39 @@ export class TestBedRender3 implements Injector, TestBed {
+      if (component) {
+        // We make a copy of the metadata to ensure that we don't mutate the original metadata
+        const metadata = {...component};
+        compileComponent(declaration, metadata);
```
We should probably store existing `ngComponentDef` here before we compile the component, similar to what we have in [compileNgModule](https://github.com/angular/angular/blob/master/packages/core/testing/src/r3_test_bed.ts#L537) function.

ocombe replied:
updated

AndrewKushnir reviewed (approved):
LGTM with one comment.

AndrewKushnir commented:
[Presubmit](http://test/OCL:228362929:BASE:228362970:1546974558449:963c6d2c)

mary-poppins commented:
You can preview dbbc2bc at https://pr27778-dbbc2bc.ngbuilds.io/.

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -309,9 +319,40 @@ export class TestBedRender3 implements Injector, TestBed {
+    });
+
+    this._createCompiler();
+    let resourceLoader: ResourceLoader = this._compiler.injector.get(ResourceLoader);
```
Sorry, one more comment. There is a proposal (from Misko) to make the `ResourceLoader` retrieval lazy, i.e. get the resource loader from injector when we need it (within `resolveComponentResources` callback call).

AndrewKushnir commented:
[Presubmit #2](https://test.corp.google.com/ui#id=OCL:228362929:BASE:228372483:1546977376596:d91745e9) with g3 deps updates.

mary-poppins commented:
You can preview f96a7c7 at https://pr27778-f96a7c7.ngbuilds.io/.

alxhub commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -495,9 +535,20 @@ export class TestBedRender3 implements Injector, TestBed {
+
+    const compilerFactory: TestingCompilerFactory =
+        this.platform.injector.get(TestingCompilerFactory);
+    this._compiler = compilerFactory.createTestingCompiler(this._compilerOptions);
```
I think this is the wrong approach. The `TestingCompiler` being created here is the View Engine JIT compiler, with all of its associated metadata extractors, template parsers, etc. We shouldn't be creating any View Engine related compiler instances in `R3TestBed`.

mary-poppins commented:
You can preview 39a66f1 at https://pr27778-39a66f1.ngbuilds.io/.

kara commented:
@ocombe Looks like there are legitimate test failures on CI. Can you take a look?

mary-poppins commented:
You can preview 363d790 at https://pr27778-363d790.ngbuilds.io/.

mary-poppins commented:
You can preview 4b5caee at https://pr27778-4b5caee.ngbuilds.io/.

mary-poppins commented:
You can preview 0629c5c at https://pr27778-0629c5c.ngbuilds.io/.

mary-poppins commented:
You can preview 5153059 at https://pr27778-5153059.ngbuilds.io/.

mary-poppins commented:
You can preview db9ef03 at https://pr27778-db9ef03.ngbuilds.io/.

ocombe committed:
feat(ivy): implement `compileComponents` method for `TestBedRender3`

The implementation of the `compileComponents` method for `TestBedRender3` was missing.
We now pass each component through `resolveComponentResources` when `TestBed.compileComponents` is called so that `templateUrl` and `styleUrls` can be resolved asynchronously and used once `TestBed.createComponent` is called.
The component's metadata are overriden in `TestBed` instead of mutating the original metadata like this is the case outside of TestBed. The reason for that is that we need to ensure that we didn't mutate anything so that the following tests can run with the same original metadata, otherwise we it could trigger or hide some errors.

FW-553 #resolve

ocombe committed:
fixup! feat(ivy): implement `compileComponents` method for `TestBedRender3`

mary-poppins commented:
You can preview b2b7ee6 at https://pr27778-b2b7ee6.ngbuilds.io/.

alxhub commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -594,7 +595,14 @@ export class TestBedRender3 implements Injector, TestBed {
+    providers.push(...this._compilerProviders);
 
     @NgModule({providers})
     class CompilerModule {
```
This is okay for now. We should really look into making this work with an `Injector` directly, and not needing to create a whole module for it.

ocombe replied:
I'll add a todo comment

alxhub reviewed (approved):
Nice!

ocombe committed:
fixup! feat(ivy): implement `compileComponents` method for `TestBedRender3`

mary-poppins commented:
You can preview 52b6ea7 at https://pr27778-52b6ea7.ngbuilds.io/.

mary-poppins commented:
You can preview 5afffbf at https://pr27778-5afffbf.ngbuilds.io/.

alxhub commented:
[Presubmit](http://test/OCL:229210121:BASE:229210571:1547491462950:9e13e62b)

mary-poppins commented:
You can preview 210d996 at https://pr27778-210d996.ngbuilds.io/.

mary-poppins commented:
You can preview 7d9cf2f at https://pr27778-7d9cf2f.ngbuilds.io/.

mary-poppins commented:
You can preview 03ecb5f at https://pr27778-03ecb5f.ngbuilds.io/.

ocombe commented:
merge assistance: I've rebased, you probably need to run a presubmit again

alxhub commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -549,6 +589,30 @@ export class TestBedRender3 implements Injector, TestBed {
+      this._compilerInjector;
+    }
+
+    const providers = [];
```
Provide a type here, this is inferred as `any[]`.

ocombe committed:
fixup! feat(ivy): implement `compileComponents` method for `TestBedRender3`

mary-poppins commented:
You can preview 3d76dca at https://pr27778-3d76dca.ngbuilds.io/.

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:229210121:BASE:229405450:1547579785201:bbd1e633)

AndrewKushnir commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:229210121:BASE:229412637:1547581942107:f0b9b3a4) (most recent one)
